fe/fe-core/src/main/java/com/starrocks/sql/optimizer/property/ValueProperty.java

```diff
@@ -0,0 +1,159 @@
+        return valueMap;
+    }
+
+    public ValueWrapper getValueWrapper(ScalarOperator scalarOperator) {
+        return valueMap.get(scalarOperator);
+    }
+    public ScalarOperator getPredicateDesc(ScalarOperator scalarOperator) {
+        return valueMap.get(scalarOperator).getPredicateDesc();
+    }
+
+    public ValueProperty projectValueProperty(Map<ColumnRefOperator, ScalarOperator> columnRefMap) {
+        Map<ScalarOperator, ValueWrapper> newValueMap = Maps.newHashMap();
+        for (Map.Entry<ColumnRefOperator, ScalarOperator> entry : columnRefMap.entrySet()) {
+            if (valueMap.containsKey(entry.getValue()) && !entry.getValue().equals(entry.getKey())) {
+                ReplaceShuttle shuttle = new ReplaceShuttle(Map.of(entry.getValue(), entry.getKey()));
+                ScalarOperator rewriteResult = shuttle.rewrite(valueMap.get(entry.getValue()).getPredicateDesc());
+                newValueMap.put(entry.getKey(), new ValueWrapper(rewriteResult));
+            }
+        }
+        ColumnRefSet outputCols = new ColumnRefSet(columnRefMap.keySet());
+        for (Map.Entry<ScalarOperator, ValueWrapper> entry : valueMap.entrySet()) {
+            if (!newValueMap.containsKey(entry.getKey()) && outputCols.containsAll(entry.getKey().getUsedColumns())) {
+                newValueMap.put(entry.getKey(), entry.getValue());
+            }
+        }
+        return new ValueProperty(newValueMap);
+    }
+
+    public ValueProperty filterValueProperty(ValueProperty filterValueProperty) {
+        Map<ScalarOperator, ValueWrapper> newValueMap = Maps.newHashMap(valueMap);
+
+        for (Map.Entry<ScalarOperator, ValueWrapper> entry : filterValueProperty.valueMap.entrySet()) {
+            if (valueMap.containsKey(entry.getKey())) {
+                newValueMap.put(entry.getKey(), valueMap.get(entry.getKey()).andValueWrapper(entry.getValue()));
+            } else {
+                newValueMap.put(entry.getKey(), entry.getValue());
+            }
+        }
+        return new ValueProperty(newValueMap);
+    }
+
+    public static ValueProperty mergeValueProperty(List<ValueProperty> valuePropertyList) {
+        Map<ScalarOperator, ValueWrapper> newValueMap = Maps.newHashMap();
+        for (ValueProperty valueProperty : valuePropertyList) {
+            for (Map.Entry<ScalarOperator, ValueWrapper> entry : valueProperty.valueMap.entrySet()) {
+                if (newValueMap.containsKey(entry.getKey())) {
+                    newValueMap.put(entry.getKey(), newValueMap.get(entry.getKey()).orValueWrapper(entry.getValue()));
+                } else {
+                    newValueMap.put(entry.getKey(), entry.getValue());
+                }
+            }
+        }
+        return new ValueProperty(newValueMap);
+    }
+
+    public static class ValueWrapper {
+        private ScalarOperator predicateDesc;
+
+        @Nullable
+        private RangeExtractor.RangeDescriptor valueDesc;
```

Comment:
only support range?

Reply:
Yes. We mainly use the range info to derive `<=, >=` predicate based on `<, >, <=, >=` on predicate. If it's a `=` on predicate, we use the predicateDesc to dervie new preidcate.